### DISCUSSION TITLE
Minor issue with voice instructions

### DISCUSSION THREAD

**Discussion by DolomitesTom** [devemux86/cruiser#164] [Q&A] [answered]
When standing still at a traffic lite of a crossing which is a turn instruction at the same time, there is the voice command "turn right in 300 meters" in an loop as long as the traffic light is red. This is a bit strange, but happens not too often, had it the first time actually. I have voice commands depending on speed active. But also the 300m were not true. I was close to the turn at the crossing.

**Comment by devemux86** [comment]
@DolomitesTom thanks for the report.

Can you post the route and the location where this happened?
(if you want privately you can send me email)

Also can you export the app settings `.pref` file?
(so I can check all navigation / voice settings)

**Comment by DolomitesTom** [comment]
which email address? I didn't save it and it is not in the cache any more

**Reply by DolomitesTom** [reply]
is ok was in the cache on the pc, email sent

**Reply by devemux86** [reply]
> which email address?

You can also use the Cruiser email mentioned on [Google Play](https://play.google.com/store/apps/details?id=gr.talent.cruiser).

**Comment by devemux86** [comment]
Thanks for the files!

There is a via point on an intersection and the next instruction "Turn right" is in **4 m**.
So there are 2 instructions in almost the same position.

I will see how we can improve such situations.

**Reply by Ciutnik** [reply]
Perhaps it would be possible to cover the same improvement in the following cases:

<img src="https://github.com/devemux86/cruiser/assets/122647782/e78bc08f-f03e-4499-87de-22c34c3f63dc" width="49%"/> <img src="https://github.com/devemux86/cruiser/assets/122647782/d5c081f4-1e4f-468e-ab3e-9afaac26abd1" width="49%"/>

It seems that the "Continue" command is redundant within 20 m of the turn command.

**Reply by devemux86** [reply]
Turn instructions are generated by the routers.
Is this BRouter?

We could skip some (continue) instructions if they are close.
But this can also lead to situations where we need them?

**Reply by Ciutnik** [reply]
Yes, this is BRouter offline. 
There is also a third case of going straight ahead. Then the "Continue" instruction is repeated twice at a distance of 20 m. Surely we can't skip any turning instructions, but the continuation? For now, I can't imagine a situation where we really need it **in conjunction with the nearby turn instruction**.

Interestingly, the fourth case, U-turn, shows a quite reasonable "U-turn" instruction:

<img src="https://github.com/devemux86/cruiser/assets/122647782/82fc220a-9168-4e17-a016-1fe2a03d70ca" width="49%"/>

But of course I understand the argument that it's better to have a few unnecessary instructions than to lack one essential one 🙂

**Reply by Ciutnik** [reply]
Especially the case when the first instruction is "Continue" (there is a forward arrow on the panel) and after 20 m the next one is "Turn" it can cause confusion!

**Reply by devemux86** [reply]
Routers often generate "Continue" turn instructions at locations where the street name changes.
However BRouter does not provide street names!

Cruiser does a post-processing on all calculated routes,
so it should be easy to make such changes if we need them.

**Reply by Ciutnik** [reply]
> it should be easy to make such changes if we need them

I vote for it. At least in the most confusing case where the "Continue" is immediately followed by "Turn".

**Reply by devemux86** [reply]
With what distance?

Also this distance may be different for driving, cycling, walking.

**Reply by Ciutnik** [reply]
> Also this distance may be different for driving, cycling, walking.

I guess probably not. 
This is mainly about intersections of dual carriageways (as in the pictures). I'll look on the maps, but the universal 25 meters will probably cover most cases.

Edit: It seems that the BRouter will skip the "Continue" command when turning left if the distance between the lanes is about 10 meters. 

<img src="https://github.com/devemux86/cruiser/assets/122647782/1378f08b-018f-46e5-a172-84cf6e4aa733" width="50%"/>

**Reply by 0709wiwiwi** [reply]
It always remains a balancing act between what is the better setting for each usage now. 
For MTB, a shorter setting (5 m) this to merge the instructions is more common and accustomed than for other uses.  The question then of course remains to merge to which turn instruction? Brouter default 40m is already virtually unusable unless you are going to cross an airport runway.

https://github.com/abrensch/brouter/issues/332#issuecomment-1150809786

**Reply by devemux86** [reply]
`turnInstructionCatchingRange` exists in BRouter's bike and walking profiles.
I can add it as option in Cruiser settings.

I do not see it in the driving profiles.
@afischerdev does it work in driving profiles too?

**Reply by afischerdev** [reply]
@devemux86 
Yes, it is working in VoiceHintProcessor for all profiles. When not defined, default 40 m is used.

**Reply by devemux86** [reply]
@afischerdev thanks for the info!

Then I will make tests in all profiles and can probably add it as option in Cruiser settings.

**Reply by afischerdev** [reply]
@devemux86 
I'm not sure if that is a good idea. This is used to merge two voice hints to one if possible and useful. E.g. left, left to u-turn-left.
There is also a [minimal value](https://github.com/abrensch/brouter/blob/03aab82e1e22306495e5807ef52e4dbf506bf280/brouter-core/src/main/java/btools/router/OsmTrack.java#L1378). Is there more info about the current sample like a BRouter-Web link?

**Reply by devemux86** [reply]
@afischerdev I mentioned it because I see this option (and other turn instructions settings) available in BRouter website.

> Is there more info about the current sample like a BRouter-Web link?

@Ciutnik ?

**Reply by Ciutnik** [reply]
There are plenty of such examples in every city. The intersection I pictured above is [here](https://brouter.de/brouter-web/#map=18/52.23631/21.08689/standard&lonlats=21.084926,52.235957;21.086953,52.236789&profile=car-eco) (if that's what you meant).

**Reply by 0709wiwiwi** [reply]
The above example
assign turnInstructionCatchingRange = 4
[Warsaw 0.2km.gpx.txt](https://github.com/devemux86/cruiser/files/12356834/Warsaw.0.2km.gpx.txt)

**Reply by devemux86** [reply]
Important is that the discussion is about possible unnecessary nearby "Continue" instructions,
not all nearby instructions. Does this flag work on all turn instructions?

**Reply by Ciutnik** [reply]
Thanks @0709wiwiwi for looking at an example, but that's exactly how the Cruiser navigates. First "Turn left" and after 17 meters "Continue". I don't understand the effect of the `turnInstructionCatchingRange` parameter on this behavior.

**Reply by devemux86** [reply]
@Ciutnik better load the profile on [BRouter website](https://brouter.de/brouter-web/) and test various `turnInstructionCatchingRange`.

Then can export the route and see the turn instructions or import the GPX into Cruiser.

See these export instructions:
- https://github.com/devemux86/cruiser/discussions/250#discussioncomment-6508754

**Reply by afischerdev** [reply]
@Ciutnik 
Thanks for the sample.

I added an issue on the [BRouter project](https://github.com/abrensch/brouter/issues/613) for this.
But I guess it needs more then a turnInstructionCatchingRange update.

**Reply by Ciutnik** [reply]
@afischerdev Thank you for your interest in the problem.
From the point of view of the navigator, [this case](https://brouter.de/brouter-web/#map=18/52.23621/21.08656/standard&lonlats=21.086335,52.235461;21.084993,52.236178&profile=car-eco) is more important, at the same intersection turn left. The first command "Continue" is confusing, and 17 meters after it is the important "Turn left", which is very easy to overlook.

Edit: I see you noticed it yourself and reported it! Thanks! 👍

**Reply by 0709wiwiwi** [reply]
@Ciutnik
[Warsaw 0.3km assign turnInstructionCatchingRange_4m .gpx.txt](https://github.com/devemux86/cruiser/files/12358708/Warsaw.0.3km.assign.turnInstructionCatchingRange_4m.gpx.txt)
[Warsaw 0.3km_assign turnInstructionCatchingRange_default.gpx.txt](https://github.com/devemux86/cruiser/files/12358709/Warsaw.0.3km_assign.turnInstructionCatchingRange_default.gpx.txt)


>Does this flag work on all turn instructions?

**Function assign turnInstructionCatchingRange.**

> This is used to merge two voice hints to one if possible and useful.

Replaces two nearby (default within 40m range) successive instructions in a single one.
This so merged single instruction than replaces the first command instruction.

Logical & usefull: 

>E.g. left & left is to be replaced by a single u-turn-left.

Unlogical.  (In the previous Brouter version the main cause of false strange instructions)

> E.g. left & right is to be replaced by what ?  Left ? Right ? Straight ?

Best is so to keep both instructions separated, without simplification.

The logical simplification left & left into a single u-turn-left.

If the lanes separation is less than 40m there is a simplification.
I prefer an assign turnInstructionCatchingRange = 15 m or less.
Not to discuss as it is like with foods and drinks = personal preference

**Continue or many Straight instructions**.

Many router engines do generate way too many straights.
If crossing a major road a straight command can be usefull.
But some generate a straight when just passing a minor side road.

The router task is to generate usefull but preferably minimal commands.
As the most optimal TTS instructions are none and so generate less TTS !

> The first command "Continue" is confusing, and 17 meters after it is the important "Turn left", which is very easy to overlook.

Advise: Cruiser Settings > Voice guidance > Advanced > Two commands at once. 
 (No idea how close the successive instructions must be, should be tested or maybe @devemux86 can answer ?)

**Reply by Ciutnik** [reply]
Referring to the case of the following commands "Continue" + "Turn left". At this particular intersection, there are separate lanes for driving straight ahead and separate lanes for turning left. Approaching the intersection, seeing and hearing the first command to drive straight ahead, you take the appropriate lane, from which you can no longer turn left. It's easy to get lost.

**Reply by devemux86** [reply]
It would be nice if routers started providing lane details in their turn instructions.

So far only OSRM seems to offer this in their API.

**Reply by 0709wiwiwi** [reply]
@Ciutnik 
Thanks for clarification.
https://brouter.de/brouter/index.html BRouter: Let's get serious about (cycle)bike routing.  Lane guiding ? 
Anyway being a little bit less generous with handing out Straight commands would **probably** help as well.

**Reply by afischerdev** [reply]
@0709wiwiwi 
> Unlogical. (In the previous Brouter version the main cause of false strange instructions)

>>    E.g. left & right is to be replaced by what ? Left ? Right ? Straight ?

> Best is so to keep both instructions separated, without simplification.

What do you say about this [sample](http://brouter.de/brouter-web/#map=18/49.97455/8.56916/osm-mapnik-german_style&lonlats=8.567633,49.974327;8.569221,49.974934&profile=car-eco)? The distance between left/right 5 m. 
Please also have a look at [GStreetView](https://www.google.com/maps/@49.9746292,8.5684799,3a,75y,64.53h,98.82t/am=t/data=!3m10!1e1!3m8!1sjq4KJqYlBronFgBrHaD75w!2e0!6s%2F%2Fgeo0.ggpht.com%2Fcbk%3Fcb_client%3Dmaps_sv.tactile%26output%3Dthumbnail%26thumb%3D2%26panoid%3Djq4KJqYlBronFgBrHaD75w%26w%3D203%26h%3D100%26yaw%3D65%26pitch%3D0%26thumbfov%3D114%26ll%3D49.974637,8.568473!7i16384!8i8192!9m2!1b1!2i21?authuser=0&entry=ttu).

**Reply by Ciutnik** [reply]
> What do you say about this sample?

It's so simple here that the combo left+right, and that's probably how navigation instructions should look like. A completely different one is left+left or right+right (=U-turn correctly proposed by BRouter). And something else is straight+right or straight+left (=right or =left), which we have recently considered. OK, I'm already dizzy... 🤪

**Reply by 0709wiwiwi** [reply]
@afischerdev  Alex. 

> What do you say about this [sample](http://brouter.de/brouter-web/#map=18/49.97455/8.56916/osm-mapnik-german_style&lonlats=8.567633,49.974327;8.569221,49.974934&profile=car-eco)? The distance between left/right 5 m.

Most routers simply generate L+R and leave the announcement task to the app.
And hardly anyone user even makes a fuss about that annoyance in my opinion.
Apps should nicely announces such turn **COMBINATION** before the maneuver.

Your example.
 
The axes of the approaching roads are quite close together. 
_I looked at it a few more times this from all sides by street view (back in DE) .  
I thought the gasse was narrower.  Straight is sure appropriate here._
 [Google street view](https://www.google.com/maps/@49.9746292,8.5684799,3a,75y,64.53h,98.82t/am=t/data=!3m10!1e1!3m8!1sjq4KJqYlBronFgBrHaD75w!2e0!6s%2F%2Fgeo0.ggpht.com%2Fcbk%3Fcb_client%3Dmaps_sv.tactile%26output%3Dthumbnail%26thumb%3D2%26panoid%3Djq4KJqYlBronFgBrHaD75w%26w%3D203%26h%3D100%26yaw%3D65%26pitch%3D0%26thumbfov%3D114%26ll%3D49.974637,8.568473!7i16384!8i8192!9m2!1b1!2i21?authuser=0&entry=ttu)

**Reply by 0709wiwiwi** [reply]
@afisherdev  

Hi Alex.
A Brouter forum item, but here you sure will find "more" testers.
I think I understand now your question why to look at your example.
I once showed many situations where it was wrong by 40m default.
I so tested those places back now and they all appear to be solved.
As I do not constantly follow up the Brouter forum, you surprised me. 
New development in silence ? Finally tackled this successfully ?

**Reply by Ciutnik** [reply]
@afischerdev 
> What do you say about this sample?

Previously I only looked at OSM and was sure that both L+R instructions are needed here. But now I looked at StreeView and indeed, "Straight" is fully justified here... 

Thus, when combining commands, the range "from-to" should be considered. So two parameters? No less than, and no more than?

**Reply by afischerdev** [reply]
@0709wiwiwi 
> New development in silence ? Finally tackled this successfully ?

I didn't close completely but I have to reduce the activity.

@Ciutnik 
We have the chance to play with the minimal values. Please see [here](https://github.com/abrensch/brouter/issues/613), that is discussable.
I also had a short look at the lanes. We only have the lanes information in the lookup table. I think here a turn:lane info is needed.

**Reply by devemux86** [reply]
> I also had a short look at the lanes. We only have the lanes information in the lookup table. I think here a turn:lane info is needed.

@afischerdev :+1:

**Comment by devemux86** [comment, marked as answer]
Should be solved in [Cruiser 3.0.30](https://play.google.com/store/apps/details?id=gr.talent.cruiser).

It also contains other navigation improvements after reports from @0709wiwiwi.

Thanks everyone for your reports! :+1:
